src/api/db/models/Image.js

```diff
@@ -482,6 +465,85 @@ export class ImageModel {
+   * onto the Project label list when necessary. Users cannot use this endpoint
+   *
+   * @param {object} input
+   * @param {object} context
+   */
+  static async createInternalLabels(input, context) {
+    console.log('ImageModel.createInternalLabels - input: ', JSON.stringify(input));
+    const operation = async ({ label }) => {
+      return await retry(async () => {
+        console.log('ImageModel.createInternalLabels - creating label: ', JSON.stringify(label));
+
+        label.type = 'ml';
+
+        // find image, create label record
+        const image = await ImageModel.queryById(label.imageId, context);
+        if (isLabelDupe(image, label)) throw new DuplicateLabelError();
+
+        const project = await ProjectModel.queryById(image.projectId);
+        const labelRecord = createLabelRecord(label, label.mlModel);
+
+        // Check if Label Exists on Project and if not, add it
+        if (!project.labels.some((l) => { return l._id === labelRecord.labelId; })) {
+          const model = await MLModelModel.queryById(labelRecord.mlModel);
+          const cats = model.categories.filter((cat) => { return cat._id === labelRecord.labelId; });
+          project.labels.push({
+            _id: labelRecord.labelId,
+            source: labelRecord.mlModel,
+            name: labelRecord.labelId,
+            // This should always be cats[0].color unless the category wasn't defined in the DB
+            // In that case assign a random color to avoid failing and losing the inference
+            color: cats.length ? cats[0].color : randomColor()
+          });
+          await project.save();
+        }
+
+        let objExists = false;
+        for (const object of image.objects) {
+          if (_.isEqual(object.bbox, label.bbox)) {
+            object.labels.unshift(labelRecord);
+            objExists = true;
+            break;
+          }
+        }
+        if (!objExists) {
+          image.objects.unshift({
+            bbox: labelRecord.bbox,
+            locked: false,
+            labels: [labelRecord]
+          });
+        }
+
+        await image.save();
+        return { image, newLabel: labelRecord };
+      }, { retries: 2 });
+    };
+
+    try {
+      for (const label of input.labels) {
+        const res = await operation({ label });
+        console.log('ImageModel.createInternalLabels - res: ', JSON.stringify(res));
+        if (label.mlModel) {
+          await handleEvent({
+            event: 'label-added',
+            label: res.newLabel,
+            image: res.image
+          }, context);
+        }
+      }
+      return { ok: true };
+    } catch (err) {
+      // if error is uncontrolled, throw new ApolloError
+      console.log(`Image.createInternalLabels() ERROR on image ${input.imageId}: ${err}`);
+      if (err instanceof ApolloError) throw err;
+      throw new ApolloError(err);
+    }
+  }
+
   static async createLabels(input, context) {
```

Comment:
I don't have super strong feelings here, but it might be worth better distinguishing between the act of _creating_ a new `Project.label` and _applying_ a label to an image. In other words maybe this action/method should be called something like `applyLabels` or `createLabelRecords`. 

Not a high priority though. What do you think?

Reply:
Hold off for now - maybe cut a separate PR for this